### PR TITLE
syslog.lua: fix parsing of RFC 5424 structured data

### PR DESCRIPTION
**Pull request by jedisct1**

### CODE REVIEW

File: modules/syslog.lua [hunk 1]
```diff
@@ -30,9 +30,11 @@ local msg_any         = octet^0
 local msg_utf8        = bom * utf8_string
 local msg             = msg_utf8 + msg_any
 local hostname        = nilvalue + printusascii^-255
-local sd_name         = printusascii^-32
+local sd_name         = (printusascii - l.S'=]" ')^-32
 local param_name      = sd_name
-local param_value     = utf8_string
+local esc             = l.P'\\'
+local param_value_esc = l.S'"\\]'
+local param_value     = ((octet - param_value_esc) + (esc * param_value_esc))^0
```
Comment (trink) [inline modules/syslog.lua]: The param_value parser would incorrectly fail on something like this: item="foo\bar"
http://tools.ietf.org/html/rfc5424#section-6.3.3
`   A backslash ('\') followed by none of the three described characters
   is considered an invalid escape sequence.  In this case, the
   backslash MUST be treated as a regular backslash and the following
   character as a regular character.  Thus, the invalid sequence MUST
   not be altered.`

File: modules/syslog.lua [hunk 2]
```diff
@@ -30,12 +36,15 @@ local msg_any         = octet^0
...
 local hostname        = nilvalue + printusascii^-255
-local sd_name         = printusascii^-32
-local param_name      = sd_name
-local param_value     = utf8_string
-local sd_param        = param_name * '="' * param_value * '"'
+local sd_name         = (printusascii - l.S'=]" ')^-32
+local param_name      = l.C(sd_name)
+local esc             = l.P'\\'
+local param_value_esc = l.S'"\\]'
+local param_value     = l.C(((octet - param_value_esc) + (esc * octet))^0) / unescape_param_value
```
Comment (trink) [inline modules/syslog.lua]: The l.C is uncessary

**Comment by trink** [comment]
Some unit tests are needed in lpeg_syslog.lua

File: modules/syslog.lua [hunk 3]
```diff
@@ -13,10 +13,16 @@ local tonumber = tonumber
 local error = error
 local type = type
 local ipairs = ipairs
+local rawset = rawset
```
Comment (trink) [inline modules/syslog.lua]: rawset is not used

File: modules/syslog.lua [hunk 4]
```diff
@@ -30,12 +36,15 @@ local msg_any         = octet^0
...
-local sd_param        = param_name * '="' * param_value * '"'
+local sd_name         = (printusascii - l.S'=]" ')^-32
+local param_name      = l.C(sd_name)
+local esc             = l.P'\\'
+local param_value_esc = l.S'"\\]'
+local param_value     = l.C(((octet - param_value_esc) + (esc * octet))^0) / unescape_param_value
 local sd_id           = sd_name
-local sd_element      = l.P"[" * sd_id * (sp * sd_param)^0 * "]"
+local sd_param        = l.Cg(param_name * '="' * param_value * '"')
+local sd_params       = (sp * l.Ct(sd_param))^0
```
Comment (trink) [inline modules/syslog.lua]: We probably want to consider how this structure will be consumed/used in Heka.

File: modules/syslog.lua (hunk 3 above)
Comment (jedisct1) [inline modules/syslog.lua]: Sorry, I forgot a commit in this PR. `rawset` is used, and the `l.C` seems necessary.

Comment (jedisct1) [inline modules/syslog.lua]: This is how I currently use it in Heka: https://github.com/jedisct1/heka/commit/29430196298b70575eb2f35323ccbf57e1a7d111

Comment (jedisct1) [inline modules/syslog.lua]: Unit test added.

Comment (trink) [inline modules/syslog.lua]: The `/` does the capture (you still need the parens)

File: src/test/lua/lpeg_syslog.lua [hunk 5]
```diff
@@ -84,6 +84,13 @@ local function kitchen_sink()
     assert(fields["$qhour"] == "01")
     assert(fields["$minute"] == "20")
     assert(fields.msg == "imklog 5.8.6, log source = /proc/kmsg started.", fields.msg)
+
```
Comment (trink) [inline src/test/lua/lpeg_syslog.lua]: This should be an independent test

File: modules/syslog.lua [hunk 6]
```diff
@@ -30,12 +36,15 @@ local msg_any         = octet^0
...
-local sd_name         = printusascii^-32
-local param_name      = sd_name
-local param_value     = utf8_string
-local sd_param        = param_name * '="' * param_value * '"'
+local sd_name         = (printusascii - l.S'=]" ')^-32
+local param_name      = l.C(sd_name)
+local esc             = l.P'\\'
+local param_value_esc = l.S'"\\]'
+local param_value     = l.C(((octet - param_value_esc) + (esc * octet))^0) / unescape_param_value
 local sd_id           = sd_name
```
Comment (trink) [inline modules/syslog.lua]: We may want to include the sd_id in the message also for easier filtering of structured data types

File: modules/syslog.lua (hunk 3 above)
Comment (jedisct1) [inline modules/syslog.lua]: Good to know, thanks!

File: modules/syslog.lua (hunk 6 above)
Comment (jedisct1) [inline modules/syslog.lua]: Makes sense. I'm going to add it.

File: modules/syslog.lua (hunk 3 above)
Comment (jedisct1) [inline modules/syslog.lua]: Extra `l.C()` removed.

Comment (jedisct1) [inline modules/syslog.lua]: This is now an independant test. Added a couple mode test cases by the way.

File: modules/syslog.lua [hunk 7]
```diff
@@ -30,12 +36,15 @@ local msg_any         = octet^0
...
-local param_value     = utf8_string
-local sd_param        = param_name * '="' * param_value * '"'
-local sd_id           = sd_name
-local sd_element      = l.P"[" * sd_id * (sp * sd_param)^0 * "]"
+local sd_name         = (printusascii - l.S'=]" ')^-32
+local param_name      = l.C(sd_name)
+local esc             = l.P'\\'
+local param_value_esc = l.S'"\\]'
+local param_value     = (((octet - param_value_esc) + (esc * octet))^0) / unescape_param_value
+local sd_id           = l.Cg(l.Cc"id" * l.C(sd_name))
```
Comment (jedisct1) [inline modules/syslog.lua]: The `sd_id` is now included in the message, and the tests have been updated appropriately.

**Comment by jedisct1** [comment]
Is there anything left you would like to be done or changed?

File: modules/syslog.lua (hunk 7 above)
Comment (trink) [inline modules/syslog.lua]: - the extra set of parens in param_value can be remove see the example in: https://github.com/mozilla-services/lua_sandbox/pull/94/files#r35908669
- A field name of 'id' in the data can clobber the sd_id (that is true of any name used here)

**Comment by jedisct1** [comment]
Extra set of parens in `param_value`  removed, names prefixes with `_`.

**Comment by jedisct1** [comment]
Anything else?

**Comment by trink** [comment]
With the tests fixed up, it looks good.  Thanks for the contribution.

**Pull request merged**